grails-app/services/org/pih/warehouse/data/ProductSynonymDataService.groovy

```diff
@@ -56,6 +60,16 @@ class ProductSynonymDataService {
                 } else {
                     command.errors.reject("Row ${index + 1}: Locale '${params['locale']}' is not a supported locale")
                 }
+
+                // Validation not to allow having more than one synonym of DISPLAY_NAME type for one locale
+                if (!productService.validateMultipleSynonymsOfType(
```

Comment:
I think you could possibly do a custom validator on the domain side. Something like this:
```
static constraints = {
    name(nullable: false, maxSize: 255)
    locale(nullable: true, validator: { val, obj ->
        if (obj.synonymTypeCode == SynonymTypeCode.DISPLAY_NAME) {
            def duplicates = obj.product?.synonyms?.findAll {Synonym synonym ->
                synonym.locale == val && synonym.synonymTypeCode == SynonymTypeCode.DISPLAY_NAME
            }
            return duplicates.size() > 1 ? "error message" : true
        }
    })
    synonymTypeCode(nullable: false)
    updatedBy(nullable: true)
    createdBy(nullable: true)
}
```
This worked for me in this way but still requires some minor improvements and some further testing if some "lazy" won't occur (so far looks good).

Reply:
Sure, I will test it out - have never seen such a solution, thanks